File: packages/w2v/reference/src/training/worker.c
#include "tokenizer.h"
#include "training.h"

#include <limits.h>
#include <math.h>
#include <stdlib.h>

Status worker_seek_to_shard(FILE *file, size_t shard_start)
{
    if (shard_start > (size_t)LONG_MAX)
    {
        return STATUS_IO_ERROR;
    }
    return fseek(file, (long)shard_start, SEEK_SET) == 0
               ? STATUS_OK
               : STATUS_IO_ERROR;
}

void worker_free(Worker *worker)
{
    if (worker->file != NULL)
    {
        fclose(worker->file);
    }
    free(worker->hidden);
    free(worker->hidden_gradient);
    *worker = (Worker){0};
}

Status worker_initialize(
    Worker *worker,
    const Trainer *trainer,
    size_t worker_id)
{
    if (worker == NULL || trainer == NULL ||
        worker_id >= trainer->config.thread_count)
    {
        return STATUS_INVALID_ARGUMENT;
    }

    *worker = (Worker){0};
    worker->worker_id = worker_id;
    worker->learning_rate = trainer->config.initial_learning_rate;
    worker->shard_start =
        trainer->corpus->byte_size /
        trainer->config.thread_count * worker_id;

    size_t dimension = trainer->config.embedding_dimension;

    worker->hidden = calloc(dimension, sizeof(*worker->hidden));
    worker->hidden_gradient = calloc(
        dimension,
        sizeof(*worker->hidden_gradient));
    if (worker->hidden == NULL || worker->hidden_gradient == NULL)
    {
        worker_free(worker);
        return STATUS_OUT_OF_MEMORY;
    }

    worker->file = fopen(trainer->corpus->path, "rb");
    if (worker->file == NULL)
    {
        worker_free(worker);
        return STATUS_IO_ERROR;
    }

    Status status = worker_seek_to_shard(
        worker->file,
        worker->shard_start);
    if (status != STATUS_OK)
    {
        worker_free(worker);
        return status;
    }

    rng_init(
        &worker->window_rng,
        derive_seed(trainer->config.root_seed, worker_id, RNG_WINDOW),
        trainer->config.rng_algorithm);
    rng_init(
        &worker->subsampling_rng,
        derive_seed(trainer->config.root_seed, worker_id, RNG_SUBSAMPLE),
        trainer->config.rng_algorithm);
    rng_init(
        &worker->negative_rng,
        derive_seed(trainer->config.root_seed, worker_id, RNG_NEGATIVE),
        trainer->config.rng_algorithm);
    return STATUS_OK;
}

static int token_should_be_discarded(
    Worker *worker,
    const Trainer *trainer,
    size_t token)
{
    if (trainer->config.subsampling_threshold <= 0)
    {
        return 0;
    }

    double frequency =
        (double)trainer->vocab->entries[token].count /
        (double)trainer->vocab->retained_token_count;
    double threshold = trainer->config.subsampling_threshold;
    double keep_probability =
        (sqrt(frequency / threshold) + 1.0) * threshold / frequency;
    double random_value = rng_uniform(&worker->subsampling_rng);

    return keep_probability < random_value;
}

Status worker_fill_sentence(
    Worker *worker,
    Trainer *trainer,
    int *finished)
{
    worker->sentence_length = 0;
    worker->sentence_position = 0;
    *finished = 0;

    char token_text[MAX_TOKEN_LENGTH];
    int at_eof = 0;
    while (worker->sentence_length < MAX_SENTENCE_LENGTH)
    {
        Status status = token_read(worker->file, token_text, &at_eof);
        if (status != STATUS_OK)
        {
            return status;
        }
        if (at_eof)
        {
            *finished = 1;
            break;
        }

        size_t token = vocab_find(trainer->vocab, token_text);
        if (token == VOCABULARY_NOT_FOUND)
        {
            continue;
        }

        worker->local_token_count++;
        worker->epoch_token_count++;
        atomic_fetch_add_explicit(
            &trainer->processed_tokens,
            1,
            memory_order_relaxed);
        if (token == 0)
        {
            break;
        }
        if (token_should_be_discarded(worker, trainer, token))
        {
            continue;
        }

        worker->sentence[worker->sentence_length] = token;
        worker->sentence_length++;
    }
    return STATUS_OK;
}

void worker_update_learning_rate(Worker *worker, const Trainer *trainer)
{
    uint64_t tokens_since_update =
        worker->local_token_count -
        worker->last_learning_rate_update_count;
    if (tokens_since_update < trainer->config.learning_rate_update_interval)
    {
        return;
    }

    uint64_t processed_tokens = atomic_load_explicit(
        &trainer->processed_tokens,
        memory_order_relaxed);
    double total_tokens =
        (double)trainer->vocab->retained_token_count *
        trainer->config.epochs;
    real learning_rate = trainer->config.initial_learning_rate *
                         (real)(1.0 -
                                (double)processed_tokens /
                                    (total_tokens + 1.0));
    real minimum = trainer->config.initial_learning_rate * 0.0001f;

    worker->learning_rate =
        learning_rate < minimum ? minimum : learning_rate;
    worker->last_learning_rate_update_count = worker->local_token_count;
}

static void train_sentence(Trainer *trainer, Worker *worker)
{
    while (worker->sentence_position < worker->sentence_length)
    {
        worker_update_learning_rate(worker, trainer);
        ModelStep step = {
            .target_token = worker->sentence[worker->sentence_position],
            .learning_rate = worker->learning_rate,
            .trainer = trainer,
            .worker = worker,
        };

        if (trainer->config.model_kind == MODEL_CBOW)
        {
            cbow_train(&step);
        }
        else
        {
            skip_gram_train(&step);
        }
        worker->sentence_position++;
    }
}

void *worker_run(void *opaque_arguments)
{
    ThreadArguments *arguments = opaque_arguments;
    Trainer *trainer = arguments->trainer;
    Worker worker;

    arguments->status = worker_initialize(
        &worker,
        trainer,
        arguments->worker_id);
    if (arguments->status != STATUS_OK)
    {
        return NULL;
    }

    for (size_t epoch = 0;
         epoch < trainer->config.epochs && arguments->status == STATUS_OK;
         epoch++)
    {
        worker.epoch_token_count = 0;
        if (epoch > 0)
        {
            clearerr(worker.file);
            arguments->status = worker_seek_to_shard(
                worker.file,
                worker.shard_start);
        }

        int finished = 0;
        while (arguments->status == STATUS_OK && !finished)
        {
            arguments->status = worker_fill_sentence(
                &worker,
                trainer,
                &finished);
            uint64_t worker_token_limit =
                trainer->vocab->retained_token_count /
                trainer->config.thread_count;
            if (worker.epoch_token_count > worker_token_limit)
            {
                finished = 1;
            }
            if (arguments->status == STATUS_OK)
            {
                if (!finished)
                {
                    train_sentence(trainer, &worker);
                }
            }
        }
    }

    worker_free(&worker);
    return NULL;
}

File: packages/w2v/src/training/worker.rs
//! Sentence filling, subsampling, and worker-local rate from the C oracle.
use super::{ModelStep, ObjectiveLoss, cbow_train, skip_gram_train};
use crate::{
    config::{MAX_SENTENCE_LENGTH, ModelKind, Real, Status},
    corpus::Tokenizer,
    random::{Rng, RngPurpose, derive_seed},
    trainer::Trainer,
};
use std::{fs::File, io::BufReader, sync::atomic::Ordering, time::Instant};

#[derive(Clone, Debug, PartialEq)]
pub struct WorkerObservation {
    pub processed_tokens: u64,
    pub learning_rate: Real,
    pub objective_loss_sum: f64,
    pub objective_loss_count: u64,
    pub elapsed_seconds: f64,
}

pub struct Worker {
    pub worker_id: usize,
    pub shard_start: usize,
    pub sentence: Vec<usize>,
    pub hidden: Vec<Real>,
    pub hidden_gradient: Vec<Real>,
    pub local_token_count: u64,
    pub epoch_token_count: u64,
    pub last_learning_rate_update_count: u64,
    pub learning_rate: Real,
    pub window_rng: Rng,
    pub subsampling_rng: Rng,
    pub negative_rng: Rng,
    pub objective_count: u64,
    pub observations: Vec<WorkerObservation>,
    tokenizer: Tokenizer<BufReader<File>>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct WorkerState {
    pub worker_id: usize,
    pub local_token_count: u64,
    pub last_learning_rate_update_count: u64,
    pub learning_rate: Real,
    pub window_rng_state: u64,
    pub subsampling_rng_state: u64,
    pub negative_rng_state: u64,
    pub objective_count: u64,
}

impl Worker {
    pub fn initialize(trainer: &Trainer, worker_id: usize) -> Result<Self, Status> {
        if worker_id >= trainer.config.thread_count {
            return Err(Status::InvalidArgument);
        }
        let shard_start = trainer.corpus.byte_size / trainer.config.thread_count * worker_id;
        let dimension = trainer.config.embedding_dimension;
        let mut sentence = Vec::new();
        sentence
            .try_reserve_exact(MAX_SENTENCE_LENGTH)
            .map_err(|_| Status::OutOfMemory)?;
        let mut hidden = Vec::new();
        let mut hidden_gradient = Vec::new();
        hidden
            .try_reserve_exact(dimension)
            .map_err(|_| Status::OutOfMemory)?;
        hidden_gradient
            .try_reserve_exact(dimension)
            .map_err(|_| Status::OutOfMemory)?;
        hidden.resize(dimension, 0.0);
        hidden_gradient.resize(dimension, 0.0);
        let rng = |purpose| {
            Rng::new(
                derive_seed(trainer.config.root_seed, worker_id, purpose),
                trainer.config.rng_algorithm,
            )
        };
        Ok(Self {
            worker_id,
            shard_start,
            sentence,
            hidden,
            hidden_gradient,
            local_token_count: 0,
            epoch_token_count: 0,
            last_learning_rate_update_count: 0,
            learning_rate: trainer.config.initial_learning_rate,
            window_rng: rng(RngPurpose::Window),
            subsampling_rng: rng(RngPurpose::Subsample),
            negative_rng: rng(RngPurpose::Negative),
            objective_count: 0,
            observations: Vec::new(),
            tokenizer: trainer.corpus.tokenizer(shard_start)?,
        })
    }

    pub fn reset_epoch(&mut self, trainer: &Trainer) -> Result<(), Status> {
        self.epoch_token_count = 0;
        self.tokenizer = trainer.corpus.tokenizer(self.shard_start)?;
        Ok(())
    }

    pub fn export_state(&self) -> WorkerState {
        WorkerState {
            worker_id: self.worker_id,
            local_token_count: self.local_token_count,
            last_learning_rate_update_count: self.last_learning_rate_update_count,
            learning_rate: self.learning_rate,
            window_rng_state: self.window_rng.state,
            subsampling_rng_state: self.subsampling_rng.state,
            negative_rng_state: self.negative_rng.state,
            objective_count: self.objective_count,
        }
    }

    pub fn restore_state(&mut self, state: &WorkerState) -> Status {
        if state.worker_id != self.worker_id
            || !state.learning_rate.is_finite()
            || state.learning_rate <= 0.0
            || state.last_learning_rate_update_count > state.local_token_count
        {
            return Status::InvalidState;
        }
        self.local_token_count = state.local_token_count;
        self.last_learning_rate_update_count = state.last_learning_rate_update_count;
        self.learning_rate = state.learning_rate;
        self.window_rng.state = state.window_rng_state;
        self.subsampling_rng.state = state.subsampling_rng_state;
        self.negative_rng.state = state.negative_rng_state;
        self.objective_count = state.objective_count;
        Status::Ok
    }

    pub fn fill_sentence(&mut self, trainer: &Trainer) -> Result<bool, Status> {
        self.sentence.clear();
        let subsampling_threshold = trainer.config.subsampling_threshold;
        let retained_token_count = trainer.vocab.retained_token_count;
        let mut finished = false;
        while self.sentence.len() < MAX_SENTENCE_LENGTH {
            let read = self.tokenizer.read_token()?;
            if read.at_eof {
                finished = true;
                break;
            }
            let Some(token) = trainer.vocab.find(&read.token) else {
                continue;
            };
            self.local_token_count = self.local_token_count.wrapping_add(1);
            self.epoch_token_count = self.epoch_token_count.wrapping_add(1);
            if token == 0 {
                break;
            }
            if subsampling_threshold > 0.0 {
                let sample = subsampling_threshold;
                let count = trainer.vocab.entries[token].count;
                let keep_probability =
                    ((count as Real / (sample * retained_token_count as Real)).sqrt() + 1.0)
                        * (sample * retained_token_count as Real)
                        / count as Real;
                if keep_probability < self.subsampling_rng.uniform() {
                    continue;
                }
            }
            self.sentence.push(token);
        }
        Ok(finished)
    }

    pub fn update_learning_rate(&mut self, trainer: &Trainer) {
        let delta = self.local_token_count - self.last_learning_rate_update_count;
        if delta <= trainer.config.learning_rate_update_interval as u64 {
            return;
        }
        let processed_before = trainer.processed_tokens.fetch_add(delta, Ordering::Relaxed);
        let processed = processed_before + delta;
        let total = trainer.vocab.retained_token_count as f64 * trainer.config.epochs as f64;
        let rate =
            trainer.config.initial_learning_rate * (1.0 - processed as f64 / (total + 1.0)) as Real;
        let minimum = trainer.config.initial_learning_rate * 0.0001;
        self.learning_rate = rate.max(minimum);
        self.last_learning_rate_update_count = self.local_token_count;
    }

    fn flush_processed_tokens(&mut self, trainer: &Trainer) {
        let delta = self.local_token_count - self.last_learning_rate_update_count;
        if delta > 0 {
            trainer.processed_tokens.fetch_add(delta, Ordering::Relaxed);
            self.last_learning_rate_update_count = self.local_token_count;
        }
    }

    pub fn train_sentence(&mut self, trainer: &Trainer, started: Instant) -> Result<(), Status> {
        for position in 0..self.sentence.len() {
            self.update_learning_rate(trainer);
            self.objective_count += 1;
            let observe = trainer.config.observation_interval > 0
                && self
                    .objective_count
                    .is_multiple_of(trainer.config.observation_interval as u64);
            let mut step = ModelStep {
                trainer,
                target_token: self.sentence[position],
                learning_rate: self.learning_rate,
                sentence: &self.sentence,
                sentence_position: position,
                hidden: &mut self.hidden,
                hidden_gradient: &mut self.hidden_gradient,
                window_rng: &mut self.window_rng,
                negative_rng: &mut self.negative_rng,
                observe_objective: observe,
            };
            let loss: ObjectiveLoss = match trainer.config.model_kind {
                ModelKind::Cbow => cbow_train(&mut step),
                ModelKind::SkipGram => skip_gram_train(&mut step),
            }?;
            if observe && loss.count > 0 {
                self.observations.push(WorkerObservation {
                    processed_tokens: trainer.processed_tokens(),
                    learning_rate: self.learning_rate,
                    objective_loss_sum: loss.sum,
                    objective_loss_count: loss.count,
                    elapsed_seconds: started.elapsed().as_secs_f64(),
                });
            }
        }
        Ok(())
    }

    pub fn run_epoch(
        &mut self,
        trainer: &Trainer,
        reset: bool,
        started: Instant,
    ) -> Result<(), Status> {
        self.observations.clear();
        if reset {
            self.reset_epoch(trainer)?;
        }
        let mut finished = false;
        while !finished {
            finished = self.fill_sentence(trainer)?;
            let limit = trainer.vocab.retained_token_count / trainer.config.thread_count as u64;
            if self.epoch_token_count > limit {
                finished = true;
            }
            if !finished {
                self.train_sentence(trainer, started)?;
            }
        }
        self.flush_processed_tokens(trainer);
        Ok(())
    }
}
